File: src/pocketcloud/scheduler/TaskHandler.php
<?php

namespace pocketcloud\scheduler;

use pocketcloud\plugin\CloudPlugin;

final class TaskHandler {

    private int $id;
    private bool $cancelled = false;
    private int $last = 0;

    public function __construct(
        private readonly Task $task,
        private int $delay,
        private readonly int $period,
        private readonly bool $repeat,
        private readonly CloudPlugin $owner
    ) {
        $this->id = mt_rand(PHP_INT_MIN, PHP_INT_MAX);
    }

    public function cancel(): void {
        if (!$this->cancelled) {
            $this->cancelled = true;
            $this->task->onCancel();
        }
    }

    public function isCancelled(): bool {
        return $this->cancelled;
    }

    public function onUpdate(int $tick): void {
        if ($this->delay > 0) {
            if (--$this->delay == 0) {
                $this->last = $tick;
                $this->task->onRun();
                if (!$this->isRepeat()) $this->cancel();
            }
            return;
        }

        if ($tick >= ($this->last + $this->period)) {
            $this->last = $tick;
            $this->task->onRun();
            if (!$this->isRepeat()) $this->cancel();
        }
    }

    public function getId(): int {
        return $this->id;
    }

    public function getTask(): Task {
        return $this->task;
    }

    public function getDelay(): int {
        return $this->delay;
    }

    public function getPeriod(): int {
        return $this->period;
    }

    public function isRepeat(): bool {
        return $this->repeat;
    }

    public function getOwner(): CloudPlugin {
        return $this->owner;
    }
}

File: cloud-core/src/main/java/de/pocketcloud/cloud/scheduler/TaskHandler.java
package de.pocketcloud.cloud.scheduler;

import de.pocketcloud.cloud.PocketCloud;
import de.pocketcloud.cloud.plugin.CloudPlugin;
import lombok.Getter;

import java.util.concurrent.ThreadLocalRandom;

@Getter
public final class TaskHandler {

    private final long id;
    private final long nextRun;

    private final Task task;
    private final int interval;
    private final boolean repeating;
    private final CloudPlugin plugin;

    private boolean cancelled = false;

    public TaskHandler(Task task, int delay, int interval, boolean repeating, CloudPlugin plugin) {
        this.task = task;
        this.interval = interval;
        this.repeating = repeating;
        this.plugin = plugin;

        this.id = ThreadLocalRandom.current().nextLong(Long.MIN_VALUE, Long.MAX_VALUE);
        this.nextRun = PocketCloud.instance().currentTick() + delay;
    }

    public void cancel() {
        if (cancelled) return;
        cancelled = true;
        try {
            task.onCancel(PocketCloud.instance().currentTick());
        } catch (Exception e) {
            plugin.getLogger().exception("Uncaught exception occurred during onCancel of task {} on tick {}", e, task.getClass().getName(), PocketCloud.instance().currentTick());
        }
    }

    public void update(long currentTick) {
        if (currentTick >= nextRun) {
            try {
                task.onRun(currentTick);
            } catch (Exception e) {
                plugin.getLogger().exception("Uncaught exception occurred during the run of task {} on tick {}", e, task.getClass().getName(), currentTick);
                this.cancel();
            } finally {
                if (!this.repeating) cancel();
            }
        }
    }
}